### PR TITLE
fix(opentelemetry lib): decode missing scope, schema_url, and resource fields

### LINKED ISSUE #24904: OpenTelemetry source: trace decode drops scope, schema_url fields
### Description

The OpenTelemetry source trace decode path (`lib/opentelemetry-proto/src/spans.rs`) drops several OTLP fields during conversion from protobuf to Vector's `TraceEvent`. This causes data loss when events are later re-encoded to OTLP (e.g., via the `opentelemetry` sink with `codec: otlp`).

### Fields dropped during trace decode

| OTLP Protobuf Field | Expected Vector Field | Status |
|---|---|---|
| `ScopeSpans.scope.name` | `scope.name` / `%opentelemetry.scope.name` | **Not decoded** |
| `ScopeSpans.scope.version` | `scope.version` / `%opentelemetry.scope.version` | **Not decoded** |
| `ScopeSpans.scope.attributes` | `scope.attributes` / `%opentelemetry.scope.attributes` | **Not decoded** |
| `ScopeSpans.scope.dropped_attributes_count` | `scope.dropped_attributes_count` / `%opentelemetry.scope.dropped_attributes_count` | **Not decoded** |
| `ScopeSpans.schema_url` | `schema_url` / `%opentelemetry.scope.schema_url` | **Not decoded** |
| `ResourceSpans.schema_url` | `schema_url` / `%opentelemetry.resource.schema_url` | **Not decoded** |

### Current behavior

In `spans.rs`, the decode path iterates over `resource_spans.scope_spans[].spans[]` but only extracts span-level fields (trace_id, span_id, name, kind, timestamps, attributes, events, links, status). The `InstrumentationScope` object on `ScopeSpans` and both `schema_url` fields are silently dropped.

### Expected behavior

These fields should be decoded into the `TraceEvent`, matching how the log decode path handles them:
- **Legacy namespace**: `scope.name`, `scope.version`, `scope.attributes`, `schema_url` at event root
- **Vector namespace**: `%opentelemetry.scope.*`, `%opentelemetry.resource.schema_url`

This mirrors [issue #15500](https://github.com/vectordotdev/vector/issues/15500) which tracks the same gap for **logs** — the log decode path at least decodes `scope.name` and `scope.version`, but traces decode nothing from the scope.

### Impact

- Round-trip OTLP → Vector → OTLP loses instrumentation scope and schema_url for traces
- The encode path (in our [PR #24621](https://github.com/vectordotdev/vector/pull/24621)) handles these fields if present, but they're never populated by the decode side
- Users working with OpenTelemetry traces through Vector cannot access scope metadata for transforms/routing

### Reproduction

1. Send traces with `InstrumentationScope` populated via OTLP gRPC/HTTP to Vector's opentelemetry source
2. Use `console` sink with `codec: json` to inspect the `TraceEvent`
3. Observe that scope fields and schema_url are absent

### Version

Verified against current `master` branch.

### PR DESCRIPTION
**Pull request by szibis**
## Summary

The OTLP decode path silently drops several protobuf fields during conversion to Vector events, causing data loss and breaking round-trip fidelity (OTLP → Vector → OTLP). This PR fixes all missing fields and adds metadata preservation for lossless OTLP metric roundtrips.

#### Before this PR

| Field | Logs | Traces | Metrics |
|---|:---:|:---:|:---:|
| `scope.name` | ✅ | ❌ | ✅ (tag) |
| `scope.version` | ✅ | ❌ | ✅ (tag) |
| `scope.attributes` | ✅ | ❌ | ✅ (tags) |
| `scope.dropped_attributes_count` | ✅ | ❌ | ❌ |
| `ScopeX.schema_url` | ❌ | ❌ | ❌ |
| `ResourceX.schema_url` | ❌ | ❌ | ❌ |
| `resource.dropped_attributes_count` | ❌ | ❌ | ❌ |

#### After this PR

| Field | Logs | Traces | Metrics |
|---|:---:|:---:|:---:|
| `scope.name` | ✅ | ✅ | ✅ (tag) |
| `scope.version` | ✅ | ✅ | ✅ (tag) |
| `scope.attributes` | ✅ | ✅ | ✅ (tags) |
| `scope.dropped_attributes_count` | ✅ | ✅ | ✅ (tag) |
| `ScopeX.schema_url` | ✅ | ✅ | ✅ (tag) |
| `ResourceX.schema_url` | ✅ | ✅ | ✅ (tag) |
| `resource.dropped_attributes_count` | ✅ | ✅ | ✅ (tag) |
| `start_time_unix_nano` (metrics) | — | — | ✅ (metadata) |
| Typed attribute preservation (metrics) | — | — | ✅ (sidecar) |

#### Field mapping

**Logs** (Legacy / Vector namespace):
- `ScopeLogs.schema_url` → `scope.schema_url` / `%opentelemetry.scope.schema_url`
- `ResourceLogs.schema_url` → `schema_url` / `%opentelemetry.resources.schema_url`
- `Resource.dropped_attributes_count` → `resource_dropped_attributes_count` / `%opentelemetry.resources.dropped_attributes_count`

**Traces** (always at event root):
- `ScopeSpans.scope.*` → `scope.name`, `scope.version`, `scope.attributes`, `scope.dropped_attributes_count`
- `ScopeSpans.schema_url` → `scope.schema_url`
- `ResourceSpans.schema_url` → `schema_url`
- `Resource.dropped_attributes_count` → `resource_dropped_attributes_count`

**Metrics** (as tags, following existing `resource.*` / `scope.*` pattern):
- `scope_dropped_attributes_count`, `scope_schema_url`, `resource_schema_url`, `resource_dropped_attributes_count`

### 2. `start_time_unix_nano` Preservation

OTLP metrics carry `start_time_unix_nano` on every data point, but Vector's metric model has only one timestamp. Previously this was silently dropped.

Now stored in `EventMetadata` at `%vector.otlp.start_time_unix_nano`:
- Non-zero values stashed via `stash_start_time()` during decode
- The encode side (PR #24897) reads this back via `resolve_start_time()`
- All 5 metric types updated (Sum, Gauge, Histogram, ExpHistogram, Summary)

### 3. Typed Metric Attribute Sidecar

OTLP attributes carry typed values (`IntValue`, `BoolValue`, `DoubleValue`), but Vector's `MetricTags` model stores everything as strings. Previously, all attributes became `StringValue` on re-encode.

Now stashed in `EventMetadata` at `%vector.otlp.metric_sidecar`:

| Sidecar Field | Content |
|---|---|
| `resource_attributes` | VRL Object preserving original types via `kv_list_into_value()` |
| `scope_attributes` | VRL Object preserving original types |
| `data_point_attributes` | VRL Object preserving original types |
| `scope_name` | String |
| `scope_version` | String |
| `resource_dropped_attributes_count` | Integer |
| `scope_dropped_attributes_count` | Integer |
| `tags_fingerprint` | Hash of stringified tags for staleness detection |

**Borrow-before-consume pattern:** The sidecar borrows `&Resource` / `&InstrumentationScope` before `build_metric_tags()` consumes them, avoiding cloning entire structures.

**Staleness detection:** The encode side (PR #24897) recomputes the fingerprint from current tags. If tags were mutated by transforms, the sidecar is ignored and the encoder falls back to string-based decomposition.

| Scenario | Before | After |
|----------|--------|-------|
| OTLP roundtrip (no transforms) | All attrs → `StringValue` | Original types preserved |
| OTLP after tag-mutating transform | `StringValue` | `StringValue` (correct fallback) |
| `start_time_unix_nano` from OTLP | Lost (hardcoded 0) | Preserved |
| Native metric (no OTLP source) | String tags | Same (backward compatible) |

### Related

- Closes #24904 (trace decode drops scope and schema_url)
- Relates to #15500 (log data model mapping gaps)
- Companion to #24621 (native log/trace → OTLP encoding)
- Companion to #24897 (native metrics → OTLP encoding, sidecar consumption)

## Test plan

- [x] 31 unit tests for missing fields (12 logs, 11 traces, 8 metrics)
- [x] 5 unit tests for `start_time_unix_nano` preservation (4 metric types + zero-not-stored)
- [x] 5 unit tests for typed sidecar (typed resource attrs, typed dp attrs, scope metadata, fingerprint validity, empty sidecar omission)
- [x] Tests verify both presence of new fields and absence when empty/zero
- [x] Tests cover Legacy and Vector namespace for logs
- [x] Combined tests verify all new fields work together with existing fields
- [ ] Integration test with OTLP collector for round-trip verification

### CODE REVIEW

**Comment by pront** [comment]
Hi @szibis, you have quite a few OTEL PRs open: https://github.com/vectordotdev/vector/pulls?q=sort%3Aupdated-desc+is%3Apr+is%3Aopen+author%3Aszibis+

Can you please list the order in which you want me to review them here? Even better, I would mark all but one as draft so I can keep filtering them without need to exchange comments here.

**Comment by szibis** [comment]
> Hi @szibis, you have quite a few OTEL PRs open: https://github.com/vectordotdev/vector/pulls?q=sort%3Aupdated-desc+is%3Apr+is%3Aopen+author%3Aszibis+
> 
> Can you please list the order in which you want me to review them here? Even better, I would mark all but one as draft so I can keep filtering them without need to exchange comments here.

@pront Sorry for that, but I just discovered this gaps and avoiding one big PR addon.

1. https://github.com/vectordotdev/vector/pull/24905 - this PR for decode missing scopes - Full OTLP format baseline for all later PR's
2. https://github.com/vectordotdev/vector/pull/24621 - For auto-convert sink in Logs and Traces
3. https://github.com/vectordotdev/vector/pull/24897 - Based on Logs and Traces native convert implement Metrics auto convert.

File: lib/opentelemetry-proto/src/metrics.rs [hunk 1]
```diff
@@ -16,30 +16,62 @@ use super::proto::{
...
-                        Some(Data::Histogram(h)) => {
-                            Self::convert_histogram(h, &resource, &scope, &metric_name)
-                        }
-                        Some(Data::ExponentialHistogram(e)) => {
-                            Self::convert_exp_histogram(e, &resource, &scope, &metric_name)
-                        }
-                        Some(Data::Summary(su)) => {
-                            Self::convert_summary(su, &resource, &scope, &metric_name)
-                        }
+                        Some(Data::Gauge(g)) => Self::convert_gauge(
```
Comment (pront) [inline lib/opentelemetry-proto/src/metrics.rs]: TIOLIT (feel free to resolve).

To avoid passing too many arguments we can introduce:

```rust
struct MetricContext<'a> {
    resource: &'a Option<Resource>,
    scope: &'a Option<InstrumentationScope>,
    metric_name: &'a str,
    scope_schema_url: &'a str,
    resource_schema_url: &'a str,
}

// usage:

let ctx = MetricContext {
    resource: &resource,
    scope: &scope,
    metric_name: &metric_name,
    scope_schema_url: &scope_schema_url,
    resource_schema_url: &resource_schema_url,
};

match metric.data {
    Some(Data::Gauge(g)) => Self::convert_gauge(g, &ctx),
    Some(Data::Sum(s)) => Self::convert_sum(s, &ctx),
    ...
}

```

File: lib/opentelemetry-proto/src/logs.rs [hunk 2]
```diff
@@ -115,18 +124,56 @@ impl ResourceLog {
...
+                path!(SCOPE_KEY, SCHEMA_URL_KEY),
+                self.scope_schema_url,
             );
         }
+
+        // Resource-level schema_url (from ResourceLogs).
+        // Legacy namespace: stored at root as "schema_url" (flat, alongside other top-level fields).
+        // Vector namespace: stored under metadata at "resources.schema_url" (grouped with resource data).
+        if !self.resource_schema_url.is_empty() {
+            log_namespace.insert_source_metadata(
```
Comment (pront) [inline lib/opentelemetry-proto/src/logs.rs]: The fix is incomplete for `LogNamespace::Vector` because the later* `resources` insert overwrites the earlier `resources.schema_url` child.

* later:
```rust
log_namespace.insert_source_metadata(
                    SOURCE_NAME,
                    &mut log,
                    Some(LegacyKey::Overwrite(path!(RESOURCE_KEY))),
                    path!(RESOURCE_KEY),
                    kv_list_into_value(resource.attributes),
                );
```

File: changelog.d/24905_otlp_decode_missing_fields.fix.md [hunk 3]
```diff
@@ -0,0 +1 @@
+The OpenTelemetry source now decodes previously dropped OTLP protobuf fields across all signal types. For logs: `ScopeLogs.schema_url`, `ResourceLogs.schema_url`, and `Resource.dropped_attributes_count` are now preserved. For traces: the entire `InstrumentationScope` (name, version, attributes, dropped_attributes_count), both `schema_url` fields, and `Resource.dropped_attributes_count` are now decoded. For metrics: `scope.dropped_attributes_count`, both `schema_url` fields, and `resource.dropped_attributes_count` are now included as metric tags. This fixes round-trip data loss when events pass through Vector between OTLP endpoints.
```
Comment (pront) [inline changelog.d/24905_otlp_decode_missing_fields.fix.md]: Nit: split into multiple lines 80-100 chars long

**Review by cswatt — DISMISSED** [review]

File: lib/opentelemetry-proto/src/spans.rs [hunk 4]
```diff
@@ -157,3 +212,305 @@ impl From<SpanStatus> for Value {
...
+                spans: vec![span],
+                schema_url: scope_schema_url.to_string(),
+            }],
+            schema_url: resource_schema_url.to_string(),
+        }
+    }
+
+    // ========================================================================
+    // Tests for InstrumentationScope decode
+    // ========================================================================
```
Comment (pront) [inline lib/opentelemetry-proto/src/spans.rs:292]: Nit:

```suggestion 

========================================================================
    // Tests for InstrumentationScope decode
========================================================================
```

here and in all other instances of this comment pattern

**Review by pront — COMMENTED** [review]
I think the new `resources.schema_url` / `resources.dropped_attributes_count` field handling introduces a backwards-compatibility issue with Vector namespace logs: they’re written into the same resources object that already contains arbitrary OTLP resource attributes, so valid incoming attributes can now be silently overwritten.

From my local test:

```json
{
  "service.name": "checkout",
  "schema_url": "tenant-defined-value",
  "dropped_attributes_count": "user-payload"
}
```

With resource metadata:

```json
{
  "schema_url": "https://resource.schema",
  "dropped_attributes_count": 7
}
```

The emitted event was:

```json
{
  "otel_resources": {
    "service.name": "checkout",
    "schema_url": "https://resource.schema",
    "dropped_attributes_count": 7
  }
}
```

So the original resource attributes `schema_url = "tenant-defined-value"` and `dropped_attributes_count = "user-payload"` were lost.

Repro config:

```yaml
data_dir: "/tmp/vector-pr-24905-data"

sources:
  otel:
    type: opentelemetry
    use_otlp_decoding: false
    log_namespace: true
    grpc:
      address: "127.0.0.1:43171"
    http:
      address: "127.0.0.1:43181"

transforms:
  expose_meta:
    type: remap
    inputs:
      - otel.logs
    source: |
      .otel_resources = %opentelemetry.resources
      .otel_scope = %opentelemetry.scope
      .otel_timestamp = %opentelemetry.timestamp

sinks:
  out:
    type: console
    inputs:
      - expose_meta
    target: stdout
    encoding:
      codec: json
```

We should preserve `%opentelemetry.resources` as the raw user-supplied resource attributes.

Also, we have the same type of issue with Metrics: `resource.*` / `scope.*` tag collisions.

Generally, if a field was not literally present as a user attribute, it should not be inserted into the raw attribute map. We should not place synthetic or derived metadata into a namespace that is also used for raw user payload.

**Comment by szibis** [comment]
> We should preserve `%opentelemetry.resources` as the raw user-supplied resource attributes.
> 
> Also, we have the same type of issue with Metrics: `resource.*` / `scope.*` tag collisions.
> 
> Generally, if a field was not literally present as a user attribute, it should not be inserted into the raw attribute map. We should not place synthetic or derived metadata into a namespace that is also used for raw user payload.

@pront All fixed

**Comment by szibis** [comment]
## Local Integration & Edge Case Testing

Set up a local test pipeline — Vector OTLP source on ports 4317/4318, piped through an OTLP sink to a second Vector instance on 4319/4320 for full roundtrip testing. Used grpcurl with proto files from `lib/opentelemetry-proto/src/proto/opentelemetry-proto/` to send payloads. Initially tried HTTP with curl but Vector's OTLP HTTP endpoint only accepts `application/x-protobuf`, not JSON, so gRPC was the only option.

That's how I caught the actual bug — passthrough output showed `time_unix_nano = 0` stored as Unix epoch (1970-01-01) instead of Null. Per OTLP spec, 0 means "timestamp not set", so this was silently corrupting data. Fixed with a `nanos_to_value()` helper that treats 0 as unset and guards against u64→i64 overflow past year 2262.

Wrote 15 unit tests in `spans.rs`, ran with `cargo test -p opentelemetry-proto --lib spans` — all green, clippy clean. Tests cover the fix plus edge cases like all SpanKind values, multiple events/links/spans, status codes, unicode, and overflow.

### Performance

Ran 10k trace events through the roundtrip pipeline in batches of 100 via grpcurl — ~5k events/s, zero errors or panics. These numbers mostly reflect grpcurl client overhead — a compiled gRPC client would be significantly faster. The important takeaway is stability: no crashes, no data corruption, no memory issues under sustained load.

**Comment by szibis** [comment]
**Note on testing methodology:** For the local integration and performance tests, all 3 PRs (#24905, #24621, #24897) were merged locally into a single test branch (`test/otlp-all-prs-merged`) and compiled into one binary. This let me test the full decode→encode→decode roundtrip across all signal types together, which is how they'll actually run in production. The unit tests in each PR are self-contained and run independently on their respective branches.

**Comment by szibis** [comment]
**#24905 (OTLP decode refactor)
      ↓
  #24621 (OTLP encode logs/traces)
      ↓
  #24897 (OTLP encode metrics)**

  Each PR rebases on top of the previous one after merge.

@pront after each merge I will do the cleanup+rebase to make this flow easiest as possible for Vector Team.

**Comment by pront** [comment]
@codex review

**Comment by pront** [comment]
Checks are failing. I recommend adding the following to your local dev env: https://github.com/vectordotdev/vector/blob/master/CONTRIBUTING.md?plain=1#L119-L160

File: lib/opentelemetry-proto/src/metrics.rs [hunk 5]
```diff
@@ -254,11 +278,195 @@ pub fn build_metric_tags(
...
+/// Stash `start_time_unix_nano` in metric metadata for OTLP roundtrip preservation.
+///
+/// Stored at `%vector.otlp.start_time_unix_nano` in EventMetadata.
+/// Only stored when non-zero (zero means "not set" in OTLP).
+/// The OTLP encoder reads this back to avoid losing start_time on roundtrip.
+fn stash_start_time(metric: &mut MetricEvent, start_time_unix_nano: u64) {
+    if start_time_unix_nano > 0 {
+        metric.metadata_mut().value_mut().insert(
+            path!("vector", "otlp", "start_time_unix_nano"),
+            Value::Integer(start_time_unix_nano as i64),
```
[the remark this answers is not in the document]
Comment (szibis) [reply lib/opentelemetry-proto/src/metrics.rs:287]: The as i64 cast is safe here - i64::MAX covers timestamps through year 2262. The bit pattern also round-trips correctly since the encode side casts back to u64. This matches chrono's timestamp_nanos(i64)  convention used throughout Vector

**Review by pront — COMMENTED** [review]
Let's also update the `tests/e2e/opentelemetry/*` tests to remap/forward the new fields and assert in `mod.rs`

**Comment by szibis** [comment]
> Let's also update the `tests/e2e/opentelemetry/*` tests to remap/forward the new fields and assert in `mod.rs`

@pront Done

**Comment by pront** [comment]
@szibis can you please resolve the merge conflicts? We aspire to merge this PR by EOW.

**Review by pront — CHANGES_REQUESTED** [review]
Hi @szibis, thanks for the thorough PR! The missing fields fix (scope, schema_url, dropped_attributes_count) is desirable and well-tested.

However, I'd like to suggest splitting this into two PRs:

* **PR 1 (bug fix)**: Decode the missing scope, schema_url, and dropped_attributes_count fields for logs, traces, and metrics. This is straightforward and can merge quickly.

* **PR 2 (new feature):** The typed metric sidecar + `start_time_unix_nano` preservation. The idea of storing a copy of the original in metadata is questionable because it creates an implicit contract between decode and encode through metadata paths and fingerprint logic. Transforms after the encoder and before the decoder can't inspect or reason about this hidden state.

I would start by exploring a potential extension of Vector's `MetricTags` model to support typed values natively, which would make the data flow through the normal pipeline without a hidden copy.

**Comment by Azyrys** [comment]
Hi Guys
Any updates on this getting close to being merged?

**Comment by pront** [comment]
> Hi Guys Any updates on this getting close to being merged?

@Azyrys per https://github.com/vectordotdev/vector/pull/24905#pullrequestreview-4064372411 there is a lot of work required to get this over the finish line.
